versioned_hdf5/backend.py

```diff
@@ -124,23 +127,143 @@ def write_dataset(f, name, data, chunks=None, dtype=None, compression=None,
+        written.
+    data_to_write : Optional[Dict[Slice, Tuple]]
+        Dict of arrays which will be written as chunks. Maps slices that will exist in
+        the raw_data once the write is complete to chunks of the dataset that is being
+        written. If ``data_to_write`` is specified, ``slices_to_write`` must be None.
+    """
+    if slices_to_write is not None and hashed_slice in slices_to_write:
+        # The hash table contains a slice we will write but haven't yet; grab the
+        # chunk from the new data being written
+        reused_chunk = new_data[slices_to_write[hashed_slice].raw]
+    elif data_to_write is not None and hashed_slice in data_to_write:
+        # The hash table contains a slice we will write but haven't yet; grab the
+        # chunk from the data_to_write dict, which stores the data that will be written
+        # for the given hashed slice.
+        reused_chunk = data_to_write[hashed_slice]
+    else:
+        # The hash table contains a slice that was written in a previous
+        # write operation; grab that chunk from the existing raw data
+        reused_slice = Tuple(
+            hashed_slice,
+            *[slice(0, size) for size in new_data.shape[1:]]
+        )
+        reused_chunk = raw_data[reused_slice.raw]
+
+    # In some cases type coersion can happen during the write process even if the dtypes
+    # are the same - for example, if the raw_data.dtype == dtype('O'), but the elements
+    # are bytes, and chunk_being_written.dtype == dtype('O'), but the elements are
+    # utf-8 strings. For this case, when the raw_data is changed, e.g.
+    #      raw_data[some_slice] = chunk_being_written[another_slice]
+    # the data that gets written is bytes. So in certain cases, just calling
+    # assert_array_equal doesn't work. Instead, we convert each element to a bytestring
+    # first.
+    if reused_chunk.dtype == 'O' and chunk_being_written.dtype == 'O':
+        to_be_written = _convert_to_bytes(chunk_being_written)
+    else:
+        to_be_written = chunk_being_written
+
+    assert_array_equal(
+        reused_chunk,
+        to_be_written,
+        err_msg=(
+            f"Hash {data_hash} of existing data chunk {reused_chunk} "
+            f"matches the hash of new data chunk {chunk_being_written}, "
+            "but data does not."
+        )
+    )
+
+def _convert_to_bytes(arr: np.ndarray) -> np.ndarray:
+    """Convert each element in the array to bytes.
+
+    Each element in the array is assumed to be the same type, even if the input is an
+    object dtype array.
+
+    Parameters
+    ----------
+    arr : np.ndarray
+        Array to be converted; no conversion is done if the elements are already bytes
+
+    Returns
+    -------
+    np.ndarray
+        Object dtype array filled with elements of type bytes
+    """
+    if len(arr) > 0 and isinstance(arr[0], bytes):
+        return arr
+    else:
+        return np.vectorize(lambda i: bytes(i, encoding='utf-8'))(arr)
```

Comment:
You can use `np.char.encode` here, I think?

Reply:
Normally I definitely would but this doesn't work for object dtype arrays:

![image](https://github.com/deshaw/versioned-hdf5/assets/14017872/cb7b3c73-95f5-4cb8-aff4-5ffd6d4c268e)